core/src/main/scala/kafka/log/TimeIndex.scala

```diff
@@ -96,20 +96,52 @@ class TimeIndex(_file: File, baseOffset: Long, maxIndexSize: Int = -1, writable:
     TimestampOffset(timestamp(buffer, n), baseOffset + relativeOffset(buffer, n))
   }
 
+  /**
+    * Convert given offset to base-relative form. This function throws InvalidOffsetException if the conversion results
+    * in overflow or underflow.
+    * @param offset Offset to convert to base-relative form.
+    * @return Base-relative offset.
+    * @throws InvalidOffsetException
+    */
+  def getRelativeOffset(offset: Long): Int = {
+    val relativeOffset = (offset - baseOffset)
+    if (relativeOffset > Integer.MAX_VALUE || relativeOffset < 0)
+      throw new InvalidOffsetException(s"Attempt to append offset $offset to time index with base offset $baseOffset will cause offset overflow")
```

Comment:
Perhaps you can add the filename to this exception message so that we'll have an easier time finding the partition if we hit this case in practice.